contracts/contracts/strategies/ConvexEthMetaStrategy.sol

```diff
@@ -257,19 +338,170 @@ contract ConvexEthMetaStrategy is InitializableAbstractStrategy {
+        // any ether that was sitting in the strategy contract before the removal.
+        uint256 ethBalance = address(this).balance;
+        // Convert all the strategy contract's ether to WETH and transfer to the vault.
+        weth.deposit{ value: ethBalance }();
+        require(
+            weth.transfer(vaultAddress, ethBalance),
+            "Transfer of WETH not successful"
+        );
+
+        emit Withdrawal(address(weth), address(lpToken), ethBalance);
+        emit Withdrawal(address(oeth), address(lpToken), oethToBurn);
+    }
+
+    /***************************************
+            Metapool Rebalancing
+    ****************************************/
+
+    /**
+     * @notice Mint OTokens and one-sided add to the Metapool.
+     * This is used when the Metapool does not have enough OTokens and too many ETH.
+     * The OToken/Asset, eg OETH/ETH, price with increase.
+     * The amount of assets in the vault is unchanged.
+     * The total supply of OTokens is increased.
+     * The asset value of the strategy and vault is increased.
+     * @param _oTokens The amount of OTokens to be minted and added to the pool.
+     */
+    function mintAndAddOTokens(uint256 _oTokens)
+        external
+        onlyStrategist
+        improveMetapoolBalance
+    {
+        IVault(vaultAddress).mintForStrategy(_oTokens);
+
+        // Convert OETH to Metapool LP tokens
+        uint256 valueInLpTokens = (_oTokens).divPrecisely(
+            curvePool.get_virtual_price()
+        );
+        // Apply slippage to LP tokens
+        uint256 minMintAmount = valueInLpTokens.mulTruncate(
+            uint256(1e18) - MAX_SLIPPAGE
+        );
+
+        // Add the minted OTokens to the Curve Metapool
+        uint256 lpDeposited = curvePool.add_liquidity(
+            [0, _oTokens],
+            minMintAmount
+        );
 
-        // Send all ETH and WETH on the contract, including extra
-        weth.deposit{ value: address(this).balance }();
+        // Deposit the Metapool LP tokens to the Convex rewards pool
         require(
-            weth.transfer(vaultAddress, weth.balanceOf(address(this))),
+            IConvexDeposits(cvxDepositorAddress).deposit(
+                cvxDepositorPTokenId,
+                lpDeposited,
+                true // Deposit with staking
+            ),
+            "Failed to Deposit LP to Convex"
+        );
+
+        emit Deposit(address(oeth), address(lpToken), _oTokens);
+    }
+
+    /**
+     * @notice One-sided remove of OTokens from the Metapool which are then burned.
+     * This is used when the Metapool has too many OTokens and not enough ETH.
+     * The amount of assets in the vault is unchanged.
+     * The total supply of OTokens is reduced.
+     * The asset value of the strategy and vault is reduced.
+     * @param _lpTokens The amount of Metapool LP tokens to be burned for OTokens.
+     */
+    function removeAndBurnOTokens(uint256 _lpTokens)
```

Comment:
We should probably go ahead and add nonreentrant to these new methods.

Reply:
The strategist would have to be compromised for a reentry attack as only the strategist can call these. Adding the reentry check adds another 2.3k gas as it needs to read the reentry flag from storage. This will be significantly cheaper if we implement transient storage (EIP-1153) that is part of the new Cancun network upgrade.

For consistency with `deposit`, `depositAll`, `withdraw` and `withdrawAll` plus the avoidance of spam bug bounties, I'll add them.

Without `nonReentrant`
```
|  Contract               ·  Method                              ·  Min        ·  Max        ·  Avg        ·  # calls      ·  eur (avg)  │
··························|······································|·············|·············|·············|···············|··············
|  ConvexEthMetaStrategy  ·  mintAndAddOTokens                   ·    1555530  ·    1559170  ·    1557350  ·            9  ·          -  │
··························|······································|·············|·············|·············|···············|··············
|  ConvexEthMetaStrategy  ·  removeAndBurnOTokens                ·    1593842  ·    1593854  ·    1593848  ·            6  ·          -  │
··························|······································|·············|·············|·············|···············|··············
|  ConvexEthMetaStrategy  ·  removeOnlyAssets                    ·    1234578  ·    1234614  ·    1234598  ·            9  ·          -  │
```

With `nonReentrant`
```
|  Contract               ·  Method                              ·  Min        ·  Max        ·  Avg        ·  # calls      ·  eur (avg)  │
··························|······································|·············|·············|·············|···············|··············
|  ConvexEthMetaStrategy  ·  mintAndAddOTokens                   ·    1557844  ·    1561484  ·    1559664  ·            9  ·          -  │
··························|······································|·············|·············|·············|···············|··············
|  ConvexEthMetaStrategy  ·  removeAndBurnOTokens                ·    1596159  ·    1596171  ·    1596165  ·            6  ·          -  │
··························|······································|·············|·············|·············|···············|··············
|  ConvexEthMetaStrategy  ·  removeOnlyAssets                    ·    1236965  ·    1237001  ·    1236985  ·            9  ·          -  │
```